tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala

```diff
@@ -11,25 +11,118 @@ import scalaz.Scalaz._
+ *   1. Operators, for which the type of the operator body is contextual
+ *
+ * An example of the first category is `\intersect`; the type of an operator intersection equals the types of both of
+ * the sets (which must have the same type). In other words, `\intersect: (Set(t), Set(t)) => Set(t)`. An example of the
+ * second is `NotSupportedByModelChecker`; its type can be anything, and is not dependent on the type of the argument
+ * (the error message).
+ *
+ * If the operator belongs to the first category, we need to implement (or extend) a `FooOperSignatures` object in
+ * [[signatures]]. Assume, for the purposes of this example, that `Foo` takes two arguments, `x: a` and `y: a -> b`, and
+ * returns a value of type `b`, for any pair of types `a` and `b`. This puts `Foo` in the first category.
+ *
+ * `foo` will have the partial signature
+ * {{{
+ *   { case Seq(a, FunT1(aa, b)) if a == aa => b }
+ * }}}
+ * as it requires exactly two arguments, such that the second argument is a function, and the domain-element type of the
+ * second argument is equal to the first argument. The return type is equal to the codomain-element type of the second
+ * argument. `FooOperSignatures` needs to return a [[SignatureMap]] (via `getMap`), containing an entry for `foo`.
+ * [[BuilderUtil.signatureMapEntry signatureMapEntry]] provides a utility method, for constructing such entries from
+ * partial functions (the implementation calls [[BuilderUtil.checkForArityException checkForArityException]] and
+ * [[BuilderUtil.completePartial completePartial]] in the background, to produce sensible error messages, if
+ * `ScopedBuilder.foo` is used with type-incorrect arguments). In our case:
+ * {{{
+ *   val fooSig = signatureMapEntry(foo, { case Seq(a, FunT1(aa, b)) if a == aa => b })
+ * }}}
+ * We then need to make sure that [[TypeComputationFactory]] adds the [[SignatureMap]] from `FooOperSignatures` to its
+ * `knownSignatures`.
+ *
+ * ==Implementing the type-safe builder method==
+ *
+ * The next step is adding (or extending) a builder in [[unsafe]], `UnsafeFooOperBuilder`. This builder should declare a
+ * method `foo(arg1: T1, ..., argN: TN): TlaEx`. The number of arguments depends on the particular operator, and is
+ * usually equal to the TLA+ arity of the operator, though this is not always the case. Typically, the argument types
+ * are `TlaEx` as well. An exception to this are, for example, operators which take type-hints. Take
+ * [[unsafe.UnsafeApalacheInternalBuilder.notSupportedByModelChecker notSupportedByModelChecker]]; it requires a
+ * `String` and a [[TlaType1]] argument, even though the TLA+ operator `NotSupportedByModelChecker` requires exactly one
+ * argument (of type `Str`).
+ *
+ * In our case, `foo` will need to accept 2 generic [[TlaEx]] arguments, so it will look like
+ * {{{
+ *   def foo(x: TlaEx, y: TlaEx): TlaEx = ...
+ * }}}
```

Comment:
To avoid any confusion with Scala generics, I suggest:

```suggestion
 * In our case, `foo` will need to accept 2 arguments of type [[TlaEx]], so it will look like
 * {{{
 *   def foo(x: TlaEx, y: TlaEx): TlaEx = ...
 * }}}
```